File: web/versions/version-utils.dart
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

part of NetTango;

class VersionUtils {

  static void updateBlocks(js.JsObject json, Function(js.JsObject) blockDefinitionHandler, Function(js.JsObject) blockInstanceHandler) {
    if (!json.hasProperty("blocks") || json["blocks"] is! js.JsArray) {
      return;
    }

    for (js.JsObject b in json["blocks"]) {
      blockDefinitionHandler(b);
    }

    if (!json.hasProperty("program") || json["program"] is! js.JsObject) {
      return;
    }

    js.JsObject program = json["program"];
    if (!program.hasProperty("chains") || program["chains"] is! js.JsArray) {
      return;
    }

    for (js.JsObject chain in program["chains"]) {
      if (chain.hasProperty("blocks") && chain["blocks"] is js.JsArray) {
        for(js.JsObject b in chain["blocks"]) {

          blockInstanceHandler(b);

          // the `children` property was removed in version 5, but this should be harmless to
          // leave in as it checks the existence of the property first.  This block can be
          // removed the next time a big structure change requires changing `updateBlocks()`.
          // -Jeremy B July 2020
          if (b.hasProperty("children") && b["children"] is js.JsArray) {
            for (js.JsObject child in b["children"]) {
              blockInstanceHandler(child);
            }
          }

          if (b.hasProperty("clauses") && b["clauses"] is js.JsArray) {
            for (js.JsObject clause in b["clauses"]) {
              if (clause.hasProperty("children") && clause["children"] is js.JsArray) {
                for (var child in clause["children"]) {
                  blockInstanceHandler(child);
                }
              }
            }
          }

        }
      }
    }

  }

  /// the data structure changed for version 4, so this remains for updating version 3 or
  /// earlier files.
  static void updateBlocks3(js.JsObject json, Function(js.JsObject) blockDefinitionHandler, Function(js.JsObject) blockInstanceHandler) {
    if (!json.hasProperty("blocks") || json["blocks"] is! js.JsArray) {
      return;
    }

    for (js.JsObject b in json["blocks"]) {
      blockDefinitionHandler(b);
    }

    // it's possible we want all the block definition updates done before processing any children
    // so run through them again.  -Jeremy B August 2019
    for (js.JsObject b in json["blocks"]) {
      if (b.hasProperty("children") && b["children"] is js.JsArray) {
        for (js.JsObject child in b["children"]) {
          blockInstanceHandler(child);
        }
      }

      if (b.hasProperty("clauses") && b["clauses"] is js.JsArray) {
        for (js.JsObject clause in b["clauses"]) {
          if (clause.hasProperty("children") && clause["children"] is js.JsArray) {
            for (js.JsObject child in clause["children"]) {
              blockInstanceHandler(child);
            }
          }
        }
      }

    }

    if (!json.hasProperty("program") || json["program"] is! js.JsObject) {
      return;
    }

    js.JsObject program = json["program"];
    if (!program.hasProperty("chains") || program["chains"] is! js.JsArray) {
      return;
    }

    for (js.JsArray bs in program["chains"]) {
      for(js.JsObject b in bs) {
        blockInstanceHandler(b);
      }
    }

  }

  static void updateBlockAttributes(js.JsObject b, Function(js.JsArray) attributesHandler) {
    if (b.hasProperty("params") && b["params"] is js.JsArray) {
      attributesHandler(b["params"]);
    }

    if (b.hasProperty("properties") && b["properties"] is js.JsArray) {
      attributesHandler(b["properties"]);
    }
  }

}

File: src/versions/version-utils.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { ArrayUtils } from "../utils/array-utils"

class VersionUtils {

  static updateBlocks(workspaceEnc: any, blockDefinitionHandler: (block: any) => void, blockInstanceHandler: (block: any) => void): void {
    workspaceEnc.blocks.forEach( (b: any) => blockDefinitionHandler(b) )
    const handleBlockInstanceRec = (b: any) => {
      blockInstanceHandler(b)
      b.clauses.forEach( (cl: any) => {
        cl.blocks.forEach( (b: any) => {
          handleBlockInstanceRec(b)
        })
      })
    }
    workspaceEnc.program.chains.forEach( (c: any) => {
      c.blocks.forEach(handleBlockInstanceRec)
    })
  }

  static updateBlocks6(json: any, blockDefinitionHandler: (block: any) => void, blockInstanceHandler: (block: Object) => void): void {
    if (!json.hasOwnProperty("blocks") || !Array.isArray(json["blocks"])) {
      return
    }

    for (var b of json["blocks"]) {
      blockDefinitionHandler(b)
    }

    if (!json.hasOwnProperty("program") || typeof(json["program"]) !== "object") {
      return
    }

    const program: any = json["program"]
    ArrayUtils.maybeForEach(program, "chains", (chain) => {
      ArrayUtils.maybeForEach(chain, "blocks", (b) => {
        blockInstanceHandler(b)
        // the `children` property was removed in version 5, but this should be harmless to
        // leave in as it checks the existence of the property first.  This line can be
        // removed the next time a big structure change requires changing `updateBlocks()`.
        // -Jeremy B July 2020
        ArrayUtils.maybeForEach(b, "children", (child) => blockInstanceHandler(child))

        ArrayUtils.maybeForEach(b, "clauses", (clause) => {
          ArrayUtils.maybeForEach(clause, "children", (child) => blockInstanceHandler(child))
        })
      })
    })

  }

  /// the data structure changed for version 4, so this remains for updating version 3 or
  /// earlier files.
  static updateBlocks3(json: any, blockDefinitionHandler: (block: any) => void, blockInstanceHandler: (block: any) => void): void {
    if (!json.hasOwnProperty("blocks") || !Array.isArray(json["blocks"])) {
      return
    }

    for (var b of json["blocks"]) {
      blockDefinitionHandler(b)
    }

    // it's possible we want all the block definition updates done before processing any children
    // so run through them again.  -Jeremy B August 2019
    for (var b of json["blocks"]) {
      ArrayUtils.maybeForEach(b, "children", (child) => blockInstanceHandler(child))
      ArrayUtils.maybeForEach(b, "clauses", (clause) => {
        ArrayUtils.maybeForEach(clause, "children", (child) => blockInstanceHandler(child))
      })
    }

    if (!json.hasOwnProperty("program") || typeof(json["program"]) !== "object") {
      return
    }

    const program = json["program"]
    ArrayUtils.maybeForEach(program, "chains", (chain) => {
      for (var b of chain) {
        blockInstanceHandler(b)
      }
    })
  }

  static updateBlockAttributes(b: any, attributesHandler: (items: any[]) => void): void {
    if (b.hasOwnProperty("params") && Array.isArray(b["params"])) {
      attributesHandler(b["params"])
    }

    if (b.hasOwnProperty("properties") && Array.isArray(b["properties"])) {
      attributesHandler(b["properties"])
    }
  }

}

export { VersionUtils }
